v4/export/sql.go

```diff
@@ -1009,3 +1056,154 @@ func GetPartitionNames(db *sql.Conn, schema, table string) (partitions []string,
 	}, "SELECT PARTITION_NAME from INFORMATION_SCHEMA.PARTITIONS WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?", schema, table)
 	return
 }
+
+// GetPartitionTableIDs get partition tableIDs through histograms.
+// SHOW STATS_HISTOGRAMS  has db_name,table_name,partition_name but doesn't have partition id
+// mysql.stats_histograms has partition_id but doesn't have db_name,table_name,partition_name
+// So we combine the results from these two sqls to get partition ids for each table
+// If UPDATE_TIME,DISTINCT_COUNT are equal, we assume these two records can represent one line.
+// If histograms are not accurate or (UPDATE_TIME,DISTINCT_COUNT) has duplicate data, it's still fine.
+// Because the possibility is low and the effect is that we will select more than one regions in one time,
+// this will not affect the correctness of the dumping data and will not affect the memory usage much.
+// This method is tricky, but no better way is found.
+// Because TiDB v3.0.0's information_schema.partition table doesn't have partition name or partition id info
+// return (dbName -> tbName -> partitionName -> partitionID, error)
+func GetPartitionTableIDs(db *sql.Conn, tables map[string]map[string]struct{}) (map[string]map[string]map[string]int64, error) {
+	const (
+		showStatsHistogramsSQL   = "SHOW STATS_HISTOGRAMS"
+		selectStatsHistogramsSQL = "SELECT TABLE_ID,FROM_UNIXTIME(VERSION DIV 262144 DIV 1000,'%Y-%m-%d %H:%i:%s') AS UPDATE_TIME,DISTINCT_COUNT FROM mysql.stats_histograms"
+	)
+	partitionIDs := make(map[string]map[string]map[string]int64, len(tables))
+	rows, err := db.QueryContext(context.Background(), showStatsHistogramsSQL)
+	if err != nil {
+		return nil, errors.Annotatef(err, "sql: %s", showStatsHistogramsSQL)
+	}
+	results, err := GetSpecifiedColumnValuesAndClose(rows, "DB_NAME", "TABLE_NAME", "PARTITION_NAME", "UPDATE_TIME", "DISTINCT_COUNT")
+	if err != nil {
+		return nil, errors.Annotatef(err, "sql: %s", showStatsHistogramsSQL)
+	}
+	type partitionInfo struct {
+		dbName, tbName, partitionName string
+	}
+	saveMap := make(map[string]map[string]partitionInfo)
+	for _, oneRow := range results {
+		dbName, tbName, partitionName, updateTime, distinctCount := oneRow[0], oneRow[1], oneRow[2], oneRow[3], oneRow[4]
+		if len(partitionName) == 0 {
+			continue
+		}
+		if tbm, ok := tables[dbName]; ok {
+			if _, ok = tbm[tbName]; ok {
+				if _, ok = saveMap[updateTime]; !ok {
+					saveMap[updateTime] = make(map[string]partitionInfo)
+				}
+				saveMap[updateTime][distinctCount] = partitionInfo{
+					dbName:        dbName,
+					tbName:        tbName,
+					partitionName: partitionName,
+				}
+			}
+		}
+	}
```

Comment:
maybe skip below logic when `saveMap` has 0 length

Reply:
Actually this ma[ is nearly impossible to have 0 length. Still add this condtition.